### PR TITLE
Deprecate static cluster metadata

### PR DESCRIPTION
**Pull request by yux0**
**What changed?**
1. Deprecate static cluster metadata
2.  Wire up cluster metadata change callback

**Why?**
To dynamically load cluster metadata and update corresponding components 

**How did you test it?**
Integration tests and unit tests

**Potential risks**
Cluster failed to load remote cluster metadata

**Is hotfix candidate?**
N/A

### CODE REVIEW

File: client/clientBean.go [hunk 1]
```diff
@@ -163,28 +178,66 @@ func (h *clientBeanImpl) SetRemoteAdminClient(
...
+				frontend.DefaultTimeout,
+				frontend.DefaultLongPollTimeout,
+			)
+			h.setRemoteFrontendClientLocked(cluster, client)
+		}
 	}
 	return client
 }
 
 func (h *clientBeanImpl) SetRemoteFrontendClient(
```
Comment (yiminc) [inline client/clientBean.go:217]: this is not used anywhere?

File: common/cluster/metadata.go [hunk 2]
```diff
@@ -212,7 +304,196 @@ func (m *metadataImpl) ClusterNameForFailoverVersion(failoverVersion int64) stri
...
+	for clusterName, info := range clusterInfo {
+		if failoverVersionIncrement <= info.InitialFailoverVersion || info.InitialFailoverVersion <= 0 {
+			panic(fmt.Sprintf(
+				"Version increment %v is smaller than initial version: %v.",
+				failoverVersionIncrement,
+				clusterInfo,
+			))
+		}
+		if len(clusterName) == 0 {
+			panic("Cluster name in all cluster names is empty")
```
Comment (yiminc) [inline common/cluster/metadata.go:453]: message is confusing. better to include more details (other fields of this clusterInfo etc).

File: service/history/historyEngine.go [hunk 3]
```diff
@@ -81,41 +82,45 @@ import (
 const (
 	conditionalRetryCount                     = 5
 	activityCancellationMsgActivityNotStarted = "ACTIVITY_ID_NOT_STARTED"
+
+	historyServiceCallbackID = common.HistoryServiceName + "-%v"
```
Comment (yiminc) [inline service/history/historyEngine.go:86]: name is too generic. Maybe clusterInfoUpdateCallbackID

File: service/history/historyEngine.go [hunk 4]
```diff
@@ -316,9 +270,12 @@ func (e *historyEngineImpl) Stop() {
 		e.visibilityProcessor.Stop()
 	}
 
+	e.clusterMetadata.UnRegisterMetadataChangeCallback(fmt.Sprintf(historyServiceCallbackID, e.shard.GetShardID()))
```
Comment (yiminc) [inline service/history/historyEngine.go:273]: I think it is worth to define a private method as getMetadataChangeCallbackID(), then you don't need to define historyServiceCallbackID.
```
func getMetadataChangeCallbackID() string {
  return fmt.Sprintf("history-service-%d", e.shard.GetShardID())
}
```

File: service/history/historyEngine.go [hunk 5]
```diff
@@ -404,6 +361,71 @@ func (e *historyEngineImpl) registerNamespaceFailoverCallback() {
...
+			defer e.replicationTaskProcessorsLock.Unlock()
+
+			for clusterName := range newClusterMetadata {
+				if clusterName == e.currentClusterName {
+					continue
+				}
+				if processor, ok := e.replicationTaskProcessors[clusterName]; ok {
+					processor.Stop()
+					delete(e.replicationTaskProcessors, clusterName)
+				}
```
Comment (yiminc) [inline service/history/historyEngine.go:378]: I don't follow the logic here. You need to check if the existing remote cluster in replicationTaskProcessors no longer present in newClusterMetadata, then you stop and delete it.

File: service/history/replicationTaskFetcher.go [hunk 6]
```diff
@@ -165,15 +152,64 @@ func (f *ReplicationTaskFetchersImpl) Stop() {
...
+			defer f.fetchersLock.Unlock()
+
+			for clusterName := range newClusterMetadata {
+				if clusterName == currentCluster {
+					continue
+				}
+				if fetcher, ok := f.fetchers[clusterName]; ok {
+					fetcher.Stop()
+					delete(f.fetchers, clusterName)
+				}
```
Comment (yiminc) [inline service/history/replicationTaskFetcher.go:206]: same problem here, should delete existing one if they no longer exists in newClusterMetadata

File: client/clientBean.go [hunk 7]
```diff
@@ -138,23 +135,41 @@ func (h *clientBeanImpl) SetMatchingClient(
...
+	h.remoteAdminClientsLock.RUnlock()
+
 	if !ok {
-		panic(fmt.Sprintf(
-			"Unknown cluster name: %v with given cluster client map: %v.",
-			cluster,
-			h.remoteAdminClients,
-		))
+		clusterInfo, clusterFound := h.clusterMetdata.GetAllClusterInfo()[cluster]
+		if !clusterFound {
```
[the remark this answers is not in the document]
Comment (yux0) [reply client/clientBean.go:154]: No. All getClient APIs are based on cluster metadata. There is one exception on history event re-application. If the cluster information is misconfig (the remote cluster B is added in cluster A but cluster A doesn't add to cluster B). The cluster info may not found and panic.

File: common/cluster/metadata.go [hunk 8]
```diff
@@ -27,14 +27,30 @@
...
+	"go.temporal.io/server/common/collection"
+	"go.temporal.io/server/common/log"
+	"go.temporal.io/server/common/log/tag"
+	"go.temporal.io/server/common/persistence"
+	"go.temporal.io/server/internal/goro"
+)
+
+const (
+	defaultClusterMetadataPageSize = 100
+	refreshInterval                = time.Minute * 5
```
[the remark this answers is not in the document]
Comment (yux0) [reply common/cluster/metadata.go:46]: Yes. We can expose a channel to do it. I can have it in a separate PR

File: common/cluster/metadata.go (hunk 2 above)
Comment (yux0) [reply common/cluster/metadata.go:453]: updated

File: service/history/historyEngine.go (hunk 3 above)
Comment (yux0) [reply service/history/historyEngine.go:86]: updated

File: service/history/historyEngine.go (hunk 5 above)
Comment (yux0) [reply service/history/historyEngine.go:378]: The metadata register update the metadata when the fields are updated 1. enabled, 2. RPCAddress 3. Initial failover version. 
Whenever we received a callback there are three cases 1. remote remote cluster 2. add new cluster. 3. refresh remote cluster. This check covers the case 1 and case 3. 
Also add some comments about it

File: service/history/replicationTaskFetcher.go (hunk 6 above)
Comment (yux0) [reply service/history/replicationTaskFetcher.go:206]: Same

File: service/history/historyEngine.go [hunk 9]
```diff
@@ -404,6 +361,71 @@ func (e *historyEngineImpl) registerNamespaceFailoverCallback() {
 	)
 }
 
+func (e *historyEngineImpl) listenToClusterMetadataChange() {
+	e.clusterMetadata.RegisterMetadataChangeCallback(
+		fmt.Sprintf(historyServiceCallbackID, e.shard.GetShardID()),
+		func(oldClusterMetadata map[string]*cluster.ClusterInformation, newClusterMetadata map[string]*cluster.ClusterInformation) {
```
[the remark this answers is not in the document]
Comment (yux0) [reply service/history/historyEngine.go]: done

File: service/history/historyEngine.go (hunk 4 above)
Comment (yux0) [reply service/history/historyEngine.go]: done

File: client/clientBean.go (hunk 1 above)
Comment (yux0) [reply client/clientBean.go]: removed the unused setter

File: client/clientBean.go (hunk 7 above)
[the remark this answers is not in the document]
Comment (yiminc) [reply client/clientBean.go:153]: shouldn't we handle that case more gracefully?

**Review by yiminc — APPROVED** [review]

**Pull request merged**